tools/block_cache_trace_analyzer.cc

```diff
@@ -0,0 +1,325 @@
+  return "";
+}
+}  // namespace
+
+BlockCacheTraceAnalyzer::BlockCacheTraceAnalyzer(
+    const std::string& trace_file_path)
+    : trace_file_path_(trace_file_path) {
+  env_ = rocksdb::Env::Default();
+}
+
+void BlockCacheTraceAnalyzer::RecordAccess(
+    const BlockCacheTraceRecord& access) {
+  ColumnFamilyStats& cf_stats = cf_stats_map_[access.cf_name];
+  SSTFileStats& file_stats = cf_stats.fd_stats_map[access.sst_fd_number];
+  file_stats.level = access.level;
+  BlockTypeStats& block_type_stats =
+      file_stats.block_type_stats_map[access.block_type];
+  BlockStats& block_stats = block_type_stats.block_stats_map[access.block_key];
+  block_stats.AddAccess(access);
+}
+
+Status BlockCacheTraceAnalyzer::Analyze() {
+  std::unique_ptr<TraceReader> trace_reader;
+  Status s =
+      NewFileTraceReader(env_, EnvOptions(), trace_file_path_, &trace_reader);
+  if (!s.ok()) {
+    return s;
+  }
+  BlockCacheTraceReader reader(std::move(trace_reader));
+  s = reader.ReadHeader(&header_);
+  if (!s.ok()) {
+    return s;
+  }
+  while (s.ok()) {
+    BlockCacheTraceRecord access;
+    s = reader.ReadAccess(&access);
+    if (!s.ok()) {
+      break;
+    }
+    RecordAccess(access);
+  }
+  return Status::OK();
+}
+
+void BlockCacheTraceAnalyzer::PrintBlockSizeStats() {
+  HistogramStat bs_stats;
+  std::map<TraceType, HistogramStat> bt_stats_map;
+  std::map<std::string, std::map<TraceType, HistogramStat>> cf_bt_stats_map;
+  for (auto const& cf_stats : cf_stats_map_) {
+    // Stats per column family.
+    const std::string& cf_name = cf_stats.first;
+    for (auto const& file_stats : cf_stats.second.fd_stats_map) {
+      // Stats per SST file.
+      for (auto const& block_type_stats :
+           file_stats.second.block_type_stats_map) {
+        // Stats per block type.
+        const TraceType type = block_type_stats.first;
+        for (auto const& block_stats :
+             block_type_stats.second.block_stats_map) {
+          // Stats per block.
+          bs_stats.Add(block_stats.second.block_size);
+          bt_stats_map[type].Add(block_stats.second.block_size);
+          cf_bt_stats_map[cf_name][type].Add(block_stats.second.block_size);
+        }
+      }
+    }
+  }
+  fprintf(stdout, "Block size stats: \n%s", bs_stats.ToString().c_str());
+  for (auto const& bt_stats : bt_stats_map) {
+    fprintf(stdout, "Block size stats for block type %s: \n%s",
+            block_type_to_string(bt_stats.first).c_str(),
+            bt_stats.second.ToString().c_str());
+  }
+  for (auto const& cf_bt_stats : cf_bt_stats_map) {
+    const std::string& cf_name = cf_bt_stats.first;
+    for (auto const& bt_stats : cf_bt_stats.second) {
+      fprintf(stdout,
+              "Block size stats for column family %s and block type %s: \n%s",
+              cf_name.c_str(), block_type_to_string(bt_stats.first).c_str(),
+              bt_stats.second.ToString().c_str());
+    }
+  }
+}
+
+void BlockCacheTraceAnalyzer::PrintAccessCountStats() {
```

Comment:
Is it OK to make this function `const`?

Reply:
Yes. I marked all Print* functions const.